File: textintel/visual/homoglyph.py
from __future__ import annotations

from textintel.normalization.confusables import CONFUSABLE_TO_LATIN, skeleton
from textintel.visual.scripts import script_name


def confusable_hits(text: str) -> list[tuple[int, str, str, str]]:
    """(index, char, script, latin_lookalike) for non-Latin letters that map to Latin."""
    hits = []
    for i, ch in enumerate(text):
        mapped = CONFUSABLE_TO_LATIN.get(ch) or CONFUSABLE_TO_LATIN.get(ch.casefold())
        scr = script_name(ch)
        if mapped and scr not in {None, "Latin", "Common", "Symbol"}:
            hits.append((i, ch, scr or "Other", mapped))
    return hits


def confusable_skeleton(text: str) -> str:
    return skeleton(text)

File: src/visual/homoglyph.rs
use crate::core::types::ConfusableCharacter;
use crate::normalization::confusables::{confusable_map, skeleton};
use crate::visual::scripts::script_name;

pub fn confusable_hits(text: &str) -> Vec<(usize, String, String, String)> {
    let map = confusable_map();
    let mut hits = Vec::new();
    for (index, ch) in text.chars().enumerate() {
        let mapped = map.get(&ch).copied();
        let script = script_name(ch);
        if let (Some(mapped), Some(script)) = (mapped, script) {
            if !matches!(script, "Latin" | "Common" | "Symbol") {
                hits.push((index, ch.to_string(), script.to_string(), mapped.to_string()));
            }
        }
    }
    hits
}

pub fn confusable_characters(text: &str) -> Vec<ConfusableCharacter> {
    confusable_hits(text)
        .into_iter()
        .map(|(index, character, script, mapped)| ConfusableCharacter {
            character,
            index,
            script,
            confusable_with: Some(mapped),
        })
        .collect()
}

pub fn confusable_skeleton(text: &str) -> String {
    skeleton(text)
}

